cs/src/core/Device/LocalMemoryDevice.cs

```diff
@@ -0,0 +1,182 @@
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace FASTER.core
+{
+    unsafe struct IORequestLocalMemory {
+        public void* srcAddress;
+        public void* dstAddress;
+        public uint bytes;
+        public DeviceIOCompletionCallback callback;
+        public object context;
+    }
+
+    /// <summary>
+    /// Local storage device
+    /// </summary>
+    public unsafe sealed class LocalMemoryDevice : StorageDeviceBase
+    {
+        readonly byte*[] ram_segments;
+        readonly GCHandle[] ram_segment_handles;
+
+        private readonly int num_segments;
+        private readonly ConcurrentQueue<IORequestLocalMemory>[] ioQueue;
+        private readonly Thread[] ioProcessors;
+        private readonly int parallelism;
+        private bool terminated;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="capacity">The maximum number of bytes this storage device can accommondate, or CAPACITY_UNSPECIFIED if there is no such limit </param>
+        /// <param name="sz_segment">The size of each segment</param>
+        /// <param name="parallelism">Number of IO processing threads</param>
+        /// <param name="sector_size">Sector size for device (default 64)</param>
+        public LocalMemoryDevice(long capacity, long sz_segment, int parallelism, uint sector_size = 64)
+            :base("/userspace/ram/storage", sector_size, capacity)
+        {
+            if (capacity == Devices.CAPACITY_UNSPECIFIED) throw new Exception("Local memory device must have a capacity!");
+            Console.WriteLine("LocalMemoryDevice: Creating a " + capacity + " sized local memory device.");
+            num_segments = (int) (capacity / sz_segment);
+            
+            ram_segments = new byte*[num_segments];
+            ram_segment_handles = new GCHandle[num_segments];
+
+            for (int i = 0; i < num_segments; i++)
+            {
+                var new_segment = new byte[sz_segment];
+                
+                ram_segment_handles[i] = GCHandle.Alloc(new_segment, GCHandleType.Pinned);
+                ram_segments[i] = (byte*)(long)ram_segment_handles[i].AddrOfPinnedObject();
+            }
+            terminated = false;
+            ioQueue = new ConcurrentQueue<IORequestLocalMemory>[parallelism];
+            this.parallelism = parallelism;
+            ioProcessors = new Thread[parallelism];
+            for (int i = 0; i != parallelism; i++)
+            {
+                var x = i;
+                ioQueue[x] = new ConcurrentQueue<IORequestLocalMemory>();
+                ioProcessors[i] = new Thread(() => this.ProcessIOQueue(ioQueue[x]));
+                ioProcessors[i].Start();
+            }
+
+            Console.WriteLine("LocalMemoryDevice: " + ram_segments.Length + " pinned in-memory segments created, each with " + sz_segment + " bytes");
+        }
+
+        private void ProcessIOQueue(ConcurrentQueue<IORequestLocalMemory> q)
```

Comment:
Superlike the idea of a fully supported memory device!

The tradeoff needed to use this device is a bit difficult to reason about: 
either add a lot of dedicated threads to minimize the queuing when reading/writing, or keep the number of dedicated IO threads low to avoid unnecessary spinning and waste of cpu%.

Is it possible to suspend the read threads until new writes show up with some basic signaling between those code paths?

Alternately, can the ConcurrentQueue be replaced with Channel<T> which has async read/write APIs? This will subject this code to the regular task scheduler, though, and cause slowdowns - maybe having a dedicated task scheduler for such hi-pri work is a good balance?

Reply:
Yes, there is a spin-wait with `Thread.Yield()` for the queue at the moment, which is sub-optimal. We could easily add a semaphore to avoid the spin, at the expense of slightly higher latency in low-load situations. We need to avoid unnecessary signaling by each enqueue, which requires some additional logic beyond what we already have in `AsyncQueue`.